### ISSUE TITLE
Incremental version of pca_lowrank

### ISSUE THREAD

**Issue by eitanrich** [pytorch/pytorch#40770]
## 🚀 Feature
An incremental version of pca_lowrank - processing a mini-batch of samples in each iteration.
Similar to sklearn.decomposition.IncrementalPCA(), but in GPU.

## Motivation

The current implementation processes all the data in one go and can run quickly only if all the data fits into the GPU memory.

## Pitch

It is sometimes needed to perform PCA on a large dataset of images. Doing it in GPU using pytorch is great, but currently limited by GPU memory. An incremental version will allow processing large datasets.

## Alternatives

Doing PCA in CPU using sklearn :(

**Comment by colesbury** [comment]
For large problems you may also want to consider randomized PCA. 

https://research.fb.com/blog/2014/09/fast-randomized-svd/
https://github.com/facebook/fbpca

**Comment by ngimel** [comment]
cc @pearu, pca_lowrank implementation author.

**Comment by vadimkantorov** [comment]
@pearu What algorithm does pca_lowrank currently use? The docs at https://pytorch.org/docs/stable/generated/torch.pca_lowrank.html have a reference to `Finding structure with randomness` without a clear note that this is indeed implemented. From the implementation PR I vaguely remember that randomized PCA was in fact not implemented, and some more advanced LOBCPG (?) was used? Should the docs be clear about what algorithm is used?

**Comment by EricAlcaide** [comment]
Ii'm interested in this as well

**Comment by dnhkng** [comment]
I needed this too, but as this request has been open since 2020, I wrote it myself:
https://github.com/dnhkng/PCAonGPU

Happy to discuss a PR with the PyTorch Team

**Comment by sirluk** [comment]
@dnhkng I had some issues with precision with your version unfortunately. I therefore reimplemented the `_incremental_mean_and_var` method from sklearn based on the paper "Algorithms for computing the sample variance: analysis and recommendations", by Chan, Golub, and LeVeque. I also extended the implementation with the option for [truncated_svd](https://pytorch.org/docs/stable/generated/torch.svd_lowrank.html) for a considerable speedup.

I published the updated version here: [https://github.com/sirluk/pytorch_incremental_pca](https://github.com/sirluk/pytorch_incremental_pca)

**Comment by vadimkantorov** [comment]
Btw maybe more useful stuff implemented in https://github.com/torchdr/torchdr by @huguesva

**Comment by ZacharyVarley** [comment]
If you can accept stochastic approximations to PCA for your application (components will be slightly different than exact ones from SVD) then I would consider a batched Oja update. With Oja, you don't have to store anything other than the current components. For more accuracy do multiple passes or fiddle with learning rate schemes. The update equation is matrix multiplication in the batched case so it is very fast on GPUs.

https://github.com/ZacharyVarley/ApproxPCATorch
